File: UKParliament.CodeTest.Web/ClientApp/src/models/person-detail.ts
import {PersonSummary} from "./person-summary";
import {PhoneNumbers} from "./phone-numbers";

export class PersonDetail extends PersonSummary {
  lineOne: string = "";
  lineTwo: string = "";
  city: string = "";
  country: string = "";
  postCode: string = "";
  phoneNumbers: PhoneNumbers[] = [];

  hasAddress = (): boolean => {
    return [this.lineOne, this.city, this.country, this.postCode].some(x => x)
  }

  validate = (): string[] => {
    let errors: string[] = [];

    if (!this.title) {
      errors.push("Person must have a title");
    }

    if (!this.name) {
      errors.push("Person must have a name");
    }

    if (!this.email || !/\S+@\S+\.\S+/.test(this.email)) {
      errors.push("Person must have a email");
    }

    if (!this.dateOfBirth || new Date(this.dateOfBirth).getTime() > new Date().getTime()) {
      errors.push("Person must have a valid date of birth");
    }

    if (this.hasAddress() && [this.lineOne, this.city, this.country, this.postCode].some(x => !x)) {
      errors.push("If a person has an address, their line one, city, country and postcode must be entered");
    }

    errors = errors.concat([...new Set(this.phoneNumbers.flatMap(x => x.validate()))]);

    if(this.phoneNumbers.length > 0 && this.phoneNumbers.filter(x => x.isPrimary).length != 1) {
      errors.push("If a person has phone numbers, they must have only one primary phone number.");
    }

    return errors;
  }
}

File: UKParliament.CodeTest.Common/Models/PersonDetail.cs
using System.Text.Json.Serialization;

namespace UKParliament.CodeTest.Models.Models;

public record PersonDetail : PersonSummary
{
    [JsonConstructor]
    public PersonDetail(
        int id,
        string title,
        string name,
        DateOnly dateOfBirth,
        string email,
        bool accountEnabled = true,
        string lineOne = null,
        string lineTwo = null,
        string city = null,
        string country = null,
        string postCode = null,
        IEnumerable<PhoneNumbers>? phoneNumbers = null) : base(id, title, name, dateOfBirth, email, accountEnabled)
    {
        LineOne = lineOne;
        LineTwo = lineTwo;
        City = city;
        Country = country;
        PostCode = postCode;
        PhoneNumbers = phoneNumbers ?? new List<PhoneNumbers>();
    }
    
    public PersonDetail(PersonSummary summary): base(summary.Id, summary.Title, summary.Name, summary.DateOfBirth, summary.Email, summary.AccountEnabled)
    {
    }
    
    public string? LineOne { get; init; }
    public string? LineTwo { get; init; }
    public string? City { get; init; }
    public string? Country { get; init; }
    public string? PostCode { get; init; }
    public IEnumerable<PhoneNumbers>? PhoneNumbers { get; init; }
}